fs2-aws-s3/src/main/scala/fs2/aws/s3/S3.scala

```diff
@@ -162,17 +202,61 @@ object S3 {
               .build()
           ).void
 
+        def createMD5: MessageDigest = MessageDigest.getInstance("MD5")
+
+        def checksumPart(chunk: Chunk[Byte]): Array[Byte] = {
+          val md = createMD5
+          chunk.foreach(byte => md.update(byte))
+          md.digest()
+        }
+
+        def updateOverallChecksum(ref: Ref[F, MessageDigest], chunk: Chunk[Byte]): F[Unit] =
+          ref.update { md =>
+            md.update(checksumPart(chunk))
+            md
+          }
+
+        def updateMaxPartNumber(ref: Ref[F, PartNumber], partNum: PartNumber): F[Unit] =
+          ref.update(current => current.max(partNum))
+
+        def formatChecksum(checksum: Array[Byte]): F[Checksum] = ApplicativeThrow[F].catchNonFatal {
+          import scala.collection.mutable
+          val sb = new mutable.StringBuilder
+          for (b <- checksum)
+            sb.append(String.format("%02x", b))
+          sb.toString()
+        }
+        def getOverallChecksum(ref: Ref[F, MessageDigest]): F[Checksum] =
+          ref.get.flatMap(md => formatChecksum(md.digest()))
+
+        def validateETag(et: Option[ETag], maxPartNumber: PartNumber, checksum: Checksum): F[Unit] =
+          multipartETagValidation match {
+            case Some(validator) =>
+              for {
+                eTag   <- ApplicativeThrow[F].fromOption(et, NoEtagReturnedByS3)
+                result <- validator.validateETag(eTag, maxPartNumber, checksum)
+              } yield result
+            case None => Applicative[F].unit
+          }
+
         in =>
-          fs2.Stream
-            .eval(initiateMultipartUpload)
-            .flatMap { uploadId =>
-              in.chunkN(chunkSizeBytes)
-                .zip(fs2.Stream.iterate(1L)(_ + 1))
-                .through(uploadPart(uploadId))
-                .fold[List[(PartETag, PartId)]](List.empty)(_ :+ _)
-                .through(completeUpload(uploadId))
-                .handleErrorWith(ex => fs2.Stream.eval(cancelUpload(uploadId) >> Sync[F].raiseError(ex)))
-            }
+          for {
+            uploadId        <- fs2.Stream.eval(initiateMultipartUpload)
+            overallChecksum <- fs2.Stream.eval(Ref.of[F, MessageDigest](createMD5))
+            maxPartNumAcc   <- fs2.Stream.eval(Ref.of[F, PartNumber](0L))
+            eTag <- in
+              .chunkN(chunkSizeBytes)
+              .zip(fs2.Stream.iterate(1L)(_ + 1))
+              .evalTap { case (chunk, _) => updateOverallChecksum(overallChecksum, chunk) }
+              .evalTap { case (_, partNumber) => updateMaxPartNumber(maxPartNumAcc, partNumber) }
+              .through(uploadPart(uploadId))
```

Comment:
Should we update the checksum AFTER successful uploadPart ?

Reply:
scratch it, it doesn't matter, we just update digest for every chank

Reply:
@semenodm I'm not sure it matters.  If we fail to upload any of them the whole Stream will error out.  Since we don't expose the ref's outside of this function I don't think anyone would be able to check the state or anything and get a wrong answer.
Can you explain why you think this?

Reply:
i have another thought, it looks like the digest is calculated sequentially with actual upload. maybe we could do it in parallel, what do you think ?

Reply:
maybe do the digest update as the sidecar for uploadPart.

Reply:
lets complete the feature and cover with test, and then think if we can improve
I want to try run it through perf tests and see the degradation level

Reply:
@semenodm Potentially I think we could do them in parallel but you have to be careful to do the overall digest in the right order according to part sequence.  We could have a MapRef[PartNumber, Digest] and then at the end sort them by part number and computer the overall md5 from the parts.  I think it's definitely more complex to do in parallel.

Reply:
ok, let me create the benchmark for S3 multipart and we will compare how this affects overall performance. if it was some regular service I wouldn't care

Reply:
Yes, piggybacking the parts digest and concat them all together as the final step.

Reply:
@semenodm take a look at it after bfe2d41